### ISSUE TITLE
Support Google Pay Passes

### ISSUE THREAD

**Issue by smrqdt** [pretix/pretix#979]
Google Pay now supports event tickets, support for this would be useful as Google Pay is available on all newer Android devices in supported countries. 

Here is the documentation: 
* https://developers.google.com/pay/passes/
* https://developers.google.com/pay/passes/guides/overview/basics/about-google-pay-api-for-passes 

Unlike Apple Google uses a centralized API where Passes are created and stored. They can be send to the customer as URL or added to the app through Javascript.

**Comment by pc-coholic** [comment]
Passes on GPay are - for now - still invite only, unfortunately.

But I have already requested access when the API was presented at last Google IO.

Right now, we are on a waitlist with a promise of getting access "soon"...

**Comment by pc-coholic** [comment]
I finally got word from Google - unfortunately, this time they refused my access-request, because they already granted someone else API-access for pretix...

Did they grant you access by any chance, @raphaelm? If not, someone else might be already working on this in secret ;-)

**Comment by pc-coholic** [comment]
Well, I also got a set of credentials and will have a whack at this :-)

Some things to consider (in my opinion - feel free to chime in :-)):
- While pkpass/Apple Wallet is pretty much completely offline and disconnected from Apple (except that you need to sign the whole thing with a certificate they provide), GPay Passes are maintained during their whole life cycle on Google servers.
- In the pretix settings, we should inform the organizer that/which data is transmitted to Google.
- Passes should not be batch-generated and uploaded but rather on demand.
- Before the pass is generated, the user should give explicit consent that his data is transmitted to Google in order to create the pass.
- pretix should query all existing objects in regular intervals to check the "hasUsers"-property and anonymize the object if it is not loaded onto a device anymore. Ideally, we would delete the object - but as far as I am aware, this is not possible as of now.
- Register to pretix' signals in order to invalidate/expire the pass if the secret changes or to update the pass if the event/product/ticket changes.

**Comment by pc-coholic** [comment]
Well, there is a working prototype that generates passes.

![ezgif-2-19dc9fdfe1](https://user-images.githubusercontent.com/654548/44312588-058c4e00-a3fb-11e8-93cf-bb91e37dc0c9.gif)

The rest of lifecycle is not yet implemented - but we're getting there ;-)

**Comment by raphaelm** [comment]
You might like this: https://github.com/pretix/pretix/commit/b6869780742bc441888f5c2e98edb3dd582830d7

**Comment by RichiH** [comment]
Just FYI, Google Pay supports passes in Germany now. Couldn't load the ticket for DENOG, but it would be super useful for 36C3.

**Comment by raphaelm** [comment]
Yes, this is not yet fully implemented, just a prototype.

I doubt that 36C3 would use it, though, since it requires passing personal data to Google.

**Comment by RichiH** [comment]
It's a user choice if they click the link in their email or not, same as with Apple Wallet.

**Comment by raphaelm** [comment]
It's not the same, Apple Wallet is a file format that, by default, does not necessarily transfer data to Apple. But yes, when time comes, @pc-coholic will finish this plugin and then it's up to the CCC team to decide whether they'll use it.

**Comment by pc-coholic** [comment]
The plugin is like 90% done - just missing some major cleanup and testing.

The issue with GPP is, that contrary to pkpass-files, there is no actual file to hand out. Instead, pretix has to send the relevant information to Google and create a link with a signed JWT payload.

I already have implemented GPP right now in a fashion that asks the user with a popup, if they really want their data to be transmitted to Google for the creation of their passes - but once the data is with Google, there is no way to remove it ever again... I included an administrative feature to "scrub" passes by overwriting any PI with random data... But that's not exactly "deleting" it, either...

So, as Raphael said: It's up to the CCC presale team to decide if they want to use the plugin (albeit I'm sure, this discussion won't be had for 36C3). But the optics of having a "Are you okay with the CCC submitting your data to Google?" popup is not exactly great ;-)

**Comment by RichiH** [comment]
Ewwww. I agree with your concerns.

FWIW, if I click a URL which ends with .pkpass, Passdroid is able to register for that data and ingest it correctly.

**Comment by raphaelm** [comment]
I've never heard of Passdroid. Pass*An*droid can handle our attachments perfectly fine.

**Comment by flawnn** [comment]
Any update on this? Google Pay fully supports it.

**Comment by raphaelm** [comment]
No, we haven't revisited this yet due to the privacy implications, and due to little interest by our customers so far

**Comment by kiancross** [comment]
Is the current work-in-progress available anywhere?

> No, we haven't revisited this yet due to the privacy implications

Can't this be left to individual organisers to determine? If organisers have the option to disable Google Wallet passes then are the privacy concerns mitigated?

**Comment by schlomo** [comment]
Yes, please let the user decide the degree of data sharing with Google, as it is their data. For those who want to use it it is a very convenient and user friendly way of storing the ticket, that also pops up at the right time and place automatically.

**Comment by pc-coholic** [comment]
We are planning to revamp the whole passbook stack in order to better integrate things like server-side push for changed passes, etc.

Apple Passbook and GPay Passes will be the first ones to profit of those coming changes.

**Comment by schlomo** [comment]
@pc-coholic thanks a lot, sounds great!

**Comment by pc-coholic** [comment]
While we're still looking into this, out persistent not doing anything has finally paid off ;-)

https://twitter.com/MishaalRahman/status/1768745883305173312

**Comment by ThiefMaster** [comment]
As someone who attended a few events using pretix, I don't think that last statement is fully correct.

Yes, you can import pkpass files into Google Wallet, but the usability is terrible: It treats ALL such passes as sensitive, and requires screen lock confirmation to display. So when attending an event where you may frequently enter and leave the ticket-only area (and thus have to scan the ticket every time), having to unlock both the phone and then confirm again to unlock the ticket on Google Walled is very annoying. Passbook apps like PassAndroid don't have this problem, but it's not as nice as having all tickets in a single place.

So I still think that native Google Wallet support (using the Google APIs to create a pass) would be a great feature!

Privacy is a good point though, but I think this can be easily solved by only triggering the Google API calls to create the pass when you click the corresponding button. There could even be a confirmation prompt so users know ticket data will be transferred to Google.

**Comment by pc-coholic** [comment]
One of our developers is currently rewriting the whole passbook/wallet integration in pretix and Google Passes will be part of that new wallet system 🎉

**Comment by ThiefMaster** [comment]
Awesome! Looking forward to hopefully having next year's animuc/Connichi tickets (I think both use pretix) in Google Wallet ;)
